### ISSUE TITLE
memory leak / segmentation fault after ft_mem_free() when using freetype .

### ISSUE THREAD

**Issue by swen100** [gtkforphp/cairo#2]
This error still exists, so i thought i give more detailed informations.

I am able to reproduce a memory leak with the following script:
```php
<?php
$fontFile = dirname(__FILE__) . "/Vera.ttf";

$surface = new \CairoImageSurface(
    CairoFormat::ARGB32,
    750,
    750
);
$context = new \CairoContext( $surface );

try {
    $fontFaceObject = new \CairoFtFontFace($fontFile);
    //$fontFaceObject = cairo_ft_font_face_create($fontFile);
    $context->setFontFace( $fontFaceObject );
    $CairoFontMatrix = new \CairoMatrix();

    $CairoFontMatrix->scale( 12, 12 );

    for( $i = 0; $i < 10; ++$i ) {
        $angle = rand(1, 360);
        $CairoFontMatrix->rotate(deg2rad($angle));
        $context->setFontMatrix( $CairoFontMatrix );

        $x = rand(0, 750);
        $y = rand(0, 750);
        $context->MoveTo( $x, $y );
        $context->textPath( "blah" );

        $context->setSourceRgba(
            rand(0,255) / 255, rand(0,255) / 255, rand(0,255) / 255, 1 - rand(0,255) / 255
        );
        $context->fill();
    }

} catch (\CairoException $exc) {
    echo "<pre>";
    print_r($exc);
    print_r($fontFaceObject);
    echo $exc->getMessage();
    echo $exc->getTraceAsString();
    die();
}

header( "Content-Type: image/png" );
$surface->writeToPng( "php://output" );

unset($context);
unset($surface);
unset($fontFaceObject);
```

**Comment by mazen** [comment]
This has been happening for us, too, even with the newer versions we get some segmentation faults.

As far as i can see, the root cause for this is in cairo_ft_font.c

```
 error = cairo_font_face_set_user_data (
                        font_face_object->font_face,
                        &font_face_object->key,
                        face,
                        (cairo_destroy_func_t) FT_Done_Face);
```

but, alas, i have no idea how to fix that

**Comment by swen100** [comment]
Hi Mazen,

"good" to hear, that i am not the only one with this problem. Meanwhile i thought that i do something wrong and i am the only unhappy guy.

At this point i still use a slightly modified 0.2 version, which does not have this problem. But, if new PHP-versions come out, what if then this old pecl-version does not work anymore? I do not hope so.

Elizabeth seems to be very busy, but i hope that she finds some time to take a look at the code again. At this point i am not able to solve this problem, i still have too less abilities in c.

Swen

**Comment by medic123de** [comment]
I found a solution working for me.
can you please check if that also solves this problem ?

https://github.com/gtkforphp/cairo/pull/15

**Comment by medic123de** [comment]
From Discussion in #15 :
ft_mem_free() is probably be related to "Debug Build"  ( configure --debug )

**Comment by auroraeosrose** [comment]
possibly - freetype is a horribly buggy library :(  we can remove that, ti's there to clear out memory properly for use with valgrind, but if it's bombing we can just take it out

**Comment by medic123de** [comment]
@auroraeosrose 
let me  see first, please ...
this is stuck so long already, some more years probably wont hurt  ;)

I also guess you need to raise minimum cairo version. 
pecl-cairo 3.2 apparently does not work with cairo 1.8  ( unit tests are failing )

**Comment by auroraeosrose** [comment]
hmmm, I'd love a new issue with the unit test failures please - I'm trying to keep semi-decent backwards compat while still adding new features with the new versions - generally the oldest version centos/redhat ships is the "lowest" version to support (oh centos/redhat, you make life angry)

**Comment by medic123de** [comment]
@auroraeosrose 
let me check if we have an unpatched RHEL 6 around.
i'll create another issue then.

back to topic:
I created a Debug Build of PHP, and it's really really bad behaving.   
It's leaking like hell - even the fonts filehandles.

That pretty much supports my suspection, that "Debug PHP" keeps a reference to all objects, and prevents proper destroyal.  
To be very precise, <b>you</b> take care of proper destroyal ( when object is destroyed ), which means the object is probably never destroyed during run.

**Comment by medic123de** [comment]
@swen100  i verified:  PHP Debug does very nasty things and is (probably) leaking like hell with libfreetype. Nevertheless I guess i have a non-crashing version.

@auroraeosrose 
PHP Debug also leaks with Cairo. I 'd like to say that PHP Debug mechanism is very much into breaking  things. ( not freeing ressources is bad in a lot of ways and scripts ).

going back to global lib_ft hadn't been an option. 
Doing a localized lib was definetly the right decision.

I improved the change, now it's also compatible with Debug mode ( but it's leaking like hell within the  same request ).

!! please review and try #15  !!

**Comment by swen100** [comment]
solved with #17
